File: kotlin/Constraints.kt
// Polyformalism Constraint Kernel — Kotlin
// =========================================
// Kotlin brings: inline classes for zero-cost type wrappers, unsigned types (UInt, ULong)
// for natural bitwise ops, extension functions for fluent APIs, and coroutines for
// parallel constraint checking. The JVM gives us battle-tested concurrent Bloom filters.

fun constraintCheck(lower: IntArray, upper: IntArray, values: IntArray): Boolean {
    require(lower.size == 16 && upper.size == 16 && values.size == 16)
    for (i in 0 until 16) {
        if (values[i] < lower[i] || values[i] > upper[i]) return false
    }
    return true
}

fun bloomMerge(dst: ULongArray, src: ULongArray) {
    val len = minOf(dst.size, src.size)
    for (i in 0 until len) {
        dst[i] = dst[i] or src[i]  // CRDT semilattice join: bitwise OR
    }
}

fun eisensteinNorm(a: Int, b: Int): Long {
    val aa = a.toLong()
    val bb = b.toLong()
    return aa * aa - aa * bb + bb * bb
}

fun bloomHash(key: Int, totalBits: Int): Int {
    var h = key and 0x7FFFFFFF
    h = h xor (h ushr 16)
    h *= -0x45d9f3b
    h = h xor (h ushr 16)
    return ((h and Int.MAX_VALUE) % totalBits)
}

fun bloomInsert(filter: ULongArray, key: Int) {
    val totalBits = filter.size * 64
    val pos = bloomHash(key, totalBits)
    filter[pos / 64] = filter[pos / 64] or (1UL shl (pos % 64))
}

fun bloomContains(filter: ULongArray, key: Int): Boolean {
    val totalBits = filter.size * 64
    val pos = bloomHash(key, totalBits)
    return (filter[pos / 64] and (1UL shl (pos % 64))) != 0UL
}

fun main() {
    println("=== Polyformalism: Kotlin ===")

    // Constraint check
    val lower = IntArray(16) { 0 }
    val upper = IntArray(16) { 100 }
    val pass  = intArrayOf(25, 30, 35, 40, 50, 60, 70, 80, 10, 20, 30, 40, 55, 65, 75, 85)
    val fail  = IntArray(16) { 200 + it }

    println("Constraint check (pass): ${constraintCheck(lower, upper, pass)}")
    println("Constraint check (fail): ${constraintCheck(lower, upper, fail)}")

    // Bloom merge
    val bloomA = ULongArray(1000)
    val bloomB = ULongArray(1000)
    bloomInsert(bloomA, 42)
    bloomInsert(bloomA, 100)
    bloomInsert(bloomA, 500)
    bloomInsert(bloomB, 500)
    bloomInsert(bloomB, 999)

    bloomMerge(bloomA, bloomB)

    println("Bloom 42 in merged:   ${bloomContains(bloomA, 42)}")
    println("Bloom 500 in merged:  ${bloomContains(bloomA, 500)}")
    println("Bloom 999 in merged:  ${bloomContains(bloomA, 999)}")
    println("Bloom 9999 in merged: ${bloomContains(bloomA, 9999)}")

    // Eisenstein norms
    println("N(3,0)  = ${eisensteinNorm(3, 0)}")
    println("N(0,1)  = ${eisensteinNorm(0, 1)}")
    println("N(2,-1) = ${eisensteinNorm(2, -1)}")
    println("N(-1,2) = ${eisensteinNorm(-1, 2)}")
    println("N(5,5)  = ${eisensteinNorm(5, 5)}")

    // Verify
    assert(constraintCheck(lower, upper, pass))
    assert(!constraintCheck(lower, upper, fail))
    assert(bloomContains(bloomA, 42))
    assert(bloomContains(bloomA, 500))
    assert(bloomContains(bloomA, 999))
    assert(!bloomContains(bloomA, 9999))
    assert(eisensteinNorm(3, 0) == 9L)
    assert(eisensteinNorm(0, 1) == 1L)
    assert(eisensteinNorm(2, -1) == 7L)
    assert(eisensteinNorm(-1, 2) == 7L)
    assert(eisensteinNorm(5, 5) == 25L)

    println("\nAll assertions passed.")
}

File: python/constraints.py
#!/usr/bin/env python3
"""Polyformalism Constraint Kernel — Python

Python is the lingua franca. Every data scientist, ML engineer, and fleet agent speaks it.
The constraint kernel in Python uses numpy for vectorized ops — the same "implicit loop"
as R/MATLAB. numpy's bitwise_or is a single C call that does the whole array at once.

What Python reveals: the constraint check is STILL one expression.
numpy makes Python an array language too.
"""

import numpy as np
from typing import List

def constraint_check(lower: np.ndarray, upper: np.ndarray, values: np.ndarray) -> bool:
    """Check if all values are within bounds. Single expression."""
    return bool(np.all((values >= lower) & (values <= upper)))

def bloom_merge(dst: np.ndarray, src: np.ndarray) -> None:
    """Bitwise OR merge in-place. CRDT semilattice join."""
    np.bitwise_or(dst, src, out=dst)

def eisenstein_norm(a: int, b: int) -> int:
    """Eisenstein integer norm: a² - ab + b²"""
    return a * a - a * b + b * b

def bloom_hash(key: int, total_bits: int) -> int:
    h = abs(key) & 0x7FFFFFFF
    h ^= h >> 16
    h = (h * 0x45d9f3b) & 0xFFFFFFFF
    h ^= h >> 16
    return h % total_bits

def bloom_insert(filter_arr: np.ndarray, key: int) -> None:
    total_bits = len(filter_arr) * 64
    pos = bloom_hash(key, total_bits)
    filter_arr[pos // 64] |= (1 << (pos % 64))

def bloom_contains(filter_arr: np.ndarray, key: int) -> bool:
    total_bits = len(filter_arr) * 64
    pos = bloom_hash(key, total_bits)
    return bool(filter_arr[pos // 64] & (1 << (pos % 64)))

def main():
    print("=== Polyformalism: Python (numpy) ===")

    # Constraint check
    lower = np.zeros(16, dtype=np.int32)
    upper = np.full(16, 100, dtype=np.int32)
    pass_vals = np.array([25, 30, 35, 40, 50, 60, 70, 80, 10, 20, 30, 40, 55, 65, 75, 85], dtype=np.int32)
    fail_vals = np.arange(200, 216, dtype=np.int32)

    print(f"Constraint check (pass): {constraint_check(lower, upper, pass_vals)}")
    print(f"Constraint check (fail): {constraint_check(lower, upper, fail_vals)}")

    # Bloom merge
    bloom_a = np.zeros(1000, dtype=np.uint64)
    bloom_b = np.zeros(1000, dtype=np.uint64)
    for k in [42, 100, 500]:
        bloom_insert(bloom_a, k)
    for k in [500, 999]:
        bloom_insert(bloom_b, k)

    bloom_merge(bloom_a, bloom_b)

    print(f"Bloom 42 in merged:   {bloom_contains(bloom_a, 42)}")
    print(f"Bloom 500 in merged:  {bloom_contains(bloom_a, 500)}")
    print(f"Bloom 999 in merged:  {bloom_contains(bloom_a, 999)}")
    print(f"Bloom 9999 in merged: {bloom_contains(bloom_a, 9999)}")

    # Eisenstein norms
    test_cases = [(3, 0), (0, 1), (2, -1), (-1, 2), (5, 5)]
    for a, b in test_cases:
        print(f"N({a},{b}) = {eisenstein_norm(a, b)}")

    # Verify
    assert constraint_check(lower, upper, pass_vals)
    assert not constraint_check(lower, upper, fail_vals)
    assert bloom_contains(bloom_a, 42)
    assert bloom_contains(bloom_a, 500)
    assert bloom_contains(bloom_a, 999)
    assert not bloom_contains(bloom_a, 9999)
    assert eisenstein_norm(3, 0) == 9
    assert eisenstein_norm(0, 1) == 1
    assert eisenstein_norm(2, -1) == 7
    assert eisenstein_norm(-1, 2) == 7
    assert eisenstein_norm(5, 5) == 25

    print("\nAll assertions passed.")

if __name__ == "__main__":
    main()
